### PR TITLE
Refactor of image accessor API for `ImageSource`

### LINKED ISSUE #524: `ImageSource.images()` can only load sequential images
`ImageSource.images()` takes mandatory `start` and `num` parameters.

However, after `start` and `num` are specified, an `indices` array is created via:
```python
 indices = np.arange(start, min(start + num, self.n), dtype=int)
```

This is then passed as a kwarg to the subclass `_images()`. 

This makes it impossible to ask for specific, nonsequential indices when calling the public method:

```python
>>> src = RelionSource("sample_relion_data.star")
>>> src.images(indices=[7,2,14,10])
TypeError: images() missing 2 required positional arguments: 'start' and 'num'
>>> src.images(0,17, indices=[7,2,14,10])
TypeError: _images() got multiple values for keyword argument 'indices'
```

The private `_images` methods of `ImageSource` subclasses, including `RelionSource`, are written in a way that allows for the `indices` to be nonsequential. The only way to ask for indices in this way is to call the private method, but we would like the public method to be able to do this.

### PR DESCRIPTION
**Pull request by chris-langfield**
Initially solves #524 

See the discussion in #680

This PR will replace #619 entirely

### CODE REVIEW

**Comment by chris-langfield** [comment]
@j-c-c Let me know if the `_images` method in the Orientation Source makes sense. Thanks!

File: src/aspire/source/simulation.py [hunk 1]
```diff
@@ -148,16 +151,15 @@ def _populate_ctf_metadata(self, filter_indices):
             filter_values,
         )
 
-    def projections(self, start=0, num=np.inf, indices=None):
+    @property
+    def projections(self):
```
Comment (garrettwrong) [inline src/aspire/source/simulation.py:156]: I think we lost the docstring for `projections`.

File: tests/test_bispectrum.py [hunk 2]
```diff
@@ -27,7 +27,7 @@ def setUp(self):
         self.src = Simulation(L=v.resolution, n=10, vols=v, dtype=v.dtype)
 
         # Original projection image to transform.
-        self.orig_img = self.src.images(0, 1)
+        self.orig_img = self.src.images[1]
```
Comment (garrettwrong) [inline tests/test_bispectrum.py]: Although it doesn't matter for this test, just so it matches before, I think this should be index 0.

File: tests/test_simulation.py [hunk 3]
```diff
@@ -100,12 +100,12 @@ def testSimulationImagesDownsample(self):
...
         # The 'images' method should be tolerant of bounds - here we ask for 1000 images starting at index 1000,
         # so we'll get back 25 images in return instead
-        images = self.sim.images(1000, 1000)
+        images = self.sim.images[1000:2000]
         self.assertTrue(images.shape, (8, 8, 25))
 
     def testSimulationImagesDownsampleShape(self):
         self.sim.downsample(6)
-        first_image = self.sim.images(0, 1)[0]
+        first_image = self.sim.images[1][0]
```
Comment (garrettwrong) [inline tests/test_simulation.py]: `.images[0]`

File: tests/test_starfile_stack.py [hunk 4]
```diff
@@ -57,7 +57,7 @@ def testMetadata(self):
 
     def testImageDownsample(self):
         self.src.downsample(16)
-        first_image = self.src.images(0, 1)[0]
+        first_image = self.src.images[1][0]
```
Comment (garrettwrong) [inline tests/test_starfile_stack.py]: `.images[0]`

File: src/aspire/source/image.py [hunk 5]
```diff
@@ -27,7 +28,48 @@
 logger = logging.getLogger(__name__)
 
 
-class ImageSource:
+class ImageAccessor:
+    """
+    Helper class for accessing images from an ImageSource as slices via the `src.images[:,:,:]` API.
```
Comment (garrettwrong) [inline src/aspire/source/image.py]: Below you check ndim=1, so we would have `start:stop:step` syntax.

File: src/aspire/source/image.py [hunk 6]
```diff
@@ -272,18 +317,18 @@ def get_metadata(self, metadata_fields, indices=None, default_value=None):
 
         return result.to_numpy().squeeze()
 
-    def _images(self, start=0, num=np.inf, indices=None):
+    def _images_super(self, indices):
```
Comment (garrettwrong) [inline src/aspire/source/image.py]: Hrmm, this wasn't really what I was thinking this method name would end up like.  Its not bad, I'm just left wondering what options we have.

I remember discussing this just because we wanted to use ABC:

```
    @abstractmethod
    def _images(self, indices):
```

That can be useful, but I feel like we shouldn't do it when detrimental to the code.  In this case we could just call it `_images` and use super....  If someone doesn't implement I suspect would crash hard anyway.  Not sure how often users will make their own ImageSources.

Maybe there is something else we can do for a "partial implementation".

If you want to keep the raise you'd get from ABC, you could do some hackery (idea, untested) to base class `_images`:

```
if type(self)._images == ImageSource._images:
    raise ImplementationError('...')
```

Then the expectation would be that concrete classes `super()._images(...)`.  Assuming my hack works then it would only trigger for the base class.

I'm probably missing other ways, will think on it a little. If nothing looks better when actually coded up, I'm fine with `_images_super` code, but maybe a name change (`_apply_image_pipeline` or something).

File: src/aspire/source/simulation.py [hunk 7]
```diff
@@ -113,6 +113,9 @@ def __init__(
             logger.info("Appending a NoiseAdder to generation pipeline")
             self.noise_adder = NoiseAdder(seed=self.seed, noise_filter=noise_filter)
 
+        self._projections_accessor = ImageAccessor(self._projections, self.n)
```
Comment (garrettwrong) [inline src/aspire/source/simulation.py]: This came out nice :)

File: tests/test_load_images.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        # ensure that we can load particles from different mrc files,
+        # nonsequentially, and end up with the ordering we asked for
+        indices = [501, 502, 503, 504, 505, 0, 1, 2, 3, 4, 729, 728, 730, 720]
+        from_src = self.src.images[indices]
+        from_mrc = self.getParticlesFromIndices(indices)
+        self.assertTrue(np.array_equal(from_src.asnumpy(), from_mrc.asnumpy()))
+
+    def testRelionSourceEveryOther(self):
+        even_indices = [i for i in range(0, self.n, 2)]
+        odd_indices = [i for i in range(1, self.n, 2)]
```
Comment (garrettwrong) [inline tests/test_load_images.py:125]: Maybe make one of these an np.array() of indices. That would hit other branch without having a duplicated test.

File: src/aspire/source/image.py [hunk 9]
```diff
@@ -27,7 +28,48 @@
...
+    def __getitem__(self, indices):
+        """
+        ImageAccessor can be indexed via Python slice object, 1-D NumPy array, list, or a single integer,
+        corresponding to the indices of the requested images. By default, slices default to a start of 0,
+        an end of self.num_imgs, and a step of 1.
+        :return: An Image object containing the requested images.
+        """
+        if isinstance(indices, (int, np.integer)):
+            indices = np.array([indices])
+        if isinstance(indices, list):
```
Comment (garrettwrong) [inline src/aspire/source/image.py]: I might like to see this one be permissive of other iterables while we are extending it. Two cases that come to mind:

range(...)  
filter(mycondition, someIterable)

I think these could be handled similarly to list for our purposes. We don't need to `next` them or anything at this point.

File: tests/test_load_images.py (hunk 8 above)
Comment (garrettwrong) [reply tests/test_load_images.py:125]: if you do add the iterable support, you might end up duping this test to cover things anyway.

This test is an example where you want to use range directly (tho you could argue `np.arange`, maybe other cases trickier).

**Review by garrettwrong — CHANGES_REQUESTED** [review]
Hey Chris, took initial look for you. This looks really great. Not sure about you guys, but I like this syntax much better!

File: src/aspire/source/image.py (hunk 9 above)
Comment (chris-langfield) [reply src/aspire/source/image.py]: So at the point where it's passed to the `fun` method, we'd just `np.array(list())` the iterable?

File: tests/test_starfile_stack.py (hunk 4 above)
Comment (chris-langfield) [reply tests/test_starfile_stack.py]: fixed

File: tests/test_simulation.py (hunk 3 above)
Comment (chris-langfield) [reply tests/test_simulation.py]: fixed

File: tests/test_bispectrum.py (hunk 2 above)
Comment (chris-langfield) [reply tests/test_bispectrum.py]: fixed thanks

File: src/aspire/source/image.py (hunk 5 above)
Comment (chris-langfield) [reply src/aspire/source/image.py]: you're referring to the docstring, right? fixed

File: src/aspire/source/image.py (hunk 9 above)
Comment (chris-langfield) [reply src/aspire/source/image.py]: or np.fromiter

File: tests/test_load_images.py (hunk 8 above)
Comment (chris-langfield) [reply tests/test_load_images.py:125]: I realize there are now two dimensions that we want to potentially test: one is the type of input whether `int`, `list`, `range` etc, and the other is the ordering of the indices whether contiguous, every other, random, etc. We also probably want to hit those for each `ImageSource` subclass where it makes sense to.. maybe I'll look at parametrizing these

File: src/aspire/source/image.py (hunk 6 above)
Comment (garrettwrong) [reply src/aspire/source/image.py]: @chris-langfield , I may have been way over thinking this... did we try having a body first?

```
class A(metaclass=abc.ABCMeta):
    @abc.abstractmethod
    def _images(self, x):
        """                                                                                                                                                                                                         
        base abstract method.                                                                                                                                                                                       
        """
        print(f"  partial implementation {x}")


class B(A):
    def _images(self, x):
        print(f"completed implementation")
        super()._images(x)

b = B()
b._images('123'). # should run completed and partial

a = A()   # should raise

```

Comment (chris-langfield) [reply src/aspire/source/image.py]: So after playing around with this idea, I realized there's a tricky part. The `super()._images()` method contains two pieces of common code:

Before querying the subclass `_images` method it checks for `cached_im`:

```python
if self._cached_im is not None:
            logger.info("Loading images from cache")
            im = Image(self._cached_im[indices, :, :])
```

But then after getting `im = self._images(indices)`, it then applies the transformations to that **image** object

```python
        else:
            im = self._images(indices)
        return self.generation_pipeline.forward(im, indices)
```
We need the `Image` object in order to apply the transforms, but we don't want to generate the `Image` before the `cached_im` check..

I guess we could break this out into two functions, one of which checks for cached, and the other which applies transforms. We could call the second one `_images` and make it an abstract method still. What do you think?

Comment (garrettwrong) [reply src/aspire/source/image.py]: We discussed trying two approaches here, one with some mild code duplication and another using some on entry and exit helper functions (like a decorator).  Since we're currently sort of backed up, I'm thinking we just move forward with the code duplication so we can move on to more interesting things.

File: src/aspire/source/image.py (hunk 9 above)
Comment (garrettwrong) [reply src/aspire/source/image.py]: np.fromiter would be pretty clean. Something like?
```
from collections import Iterable
...
if isinstance(indices, (int, np.integer)):
   ....
elif isinstance(indices, Iterable) and not isinstance(indices, ndarray):
   indices = np.fromiter(indices, dtype=int)
elif isinstance(indices, slice):
    ....

# your existing array and dim checks
...
```

Comment (chris-langfield) [reply src/aspire/source/image.py]: The only problem here is `dict`s and `str`s are also iterable... I guess we have to pick whether to explicitly code which iterables are allowed vs which aren't?

Comment (garrettwrong) [reply src/aspire/source/image.py]: We're way over thinking this.  If someone passes a string to this code, the code is definitely going to crash soon, and that is their problem.

If you would rather handle tuples, lists, arrays, etc separately, that would be fine. Doesn't matter to me.

Comment (garrettwrong) [reply src/aspire/source/image.py]:
```
>>> x = np.fromiter('abc', dtype=int)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid literal for int() with base 10: 'a'
>>> x = np.fromiter({'k':'v'}, dtype=int)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid literal for int() with base 10: 'k'
```

Comment (garrettwrong) [reply src/aspire/source/image.py]: They won't get very far...

**Comment by chris-langfield** [comment]
Pushed cache check and forward transformations down to subclass `_image` methods. `ImageSource._images` remains an abstract method. Also fixed up key handling and fleshed out tests. Also docstrings

File: src/aspire/source/image.py [hunk 10]
```diff
@@ -27,7 +29,50 @@
...
+    def __getitem__(self, indices):
+        """
+        ImageAccessor can be indexed via Python slice object, 1-D NumPy array, list, or a single integer,
+        corresponding to the indices of the requested images. By default, slices default to a start of 0,
+        an end of self.num_imgs, and a step of 1.
+        :return: An Image object containing the requested images.
+        """
+        if isinstance(indices, Iterable) and not isinstance(indices, np.ndarray):
+            indices = np.fromiter(indices, int)
+        if isinstance(indices, (int, np.integer)):
```
Comment (garrettwrong) [inline src/aspire/source/image.py]: elif

Not for performance or anything, it just helps document intent in the code. (In case someone else needs to refactor it later).

File: src/aspire/source/image.py [hunk 11]
```diff
@@ -27,7 +29,50 @@
...
+            indices = np.array([indices])
+        if isinstance(indices, slice):
```
Comment (garrettwrong) [inline src/aspire/source/image.py]: elif

File: src/aspire/source/image.py [hunk 12]
```diff
@@ -88,6 +133,9 @@ def __init__(self, L, n, dtype="double", metadata=None, memory=None):
         self.generation_pipeline = Pipeline(xforms=None, memory=memory)
         self._metadata_out = None
 
+        # instantiate the accessor for the `images` property
```
Comment (garrettwrong) [inline src/aspire/source/image.py]: Feel weird asking you to capitalize this, but I figure someone else will if I don't :)

File: tests/test_load_images.py [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+        from_mrc = self.getParticlesFromIndices([1, 2, 3, 4, 5])
+        self.assertTrue(np.array_equal(from_src.asnumpy(), from_mrc.asnumpy()))
+
+    def testBadInput(self):
+        # test a non-iterable, non-NumPy, non-integer input
+        with self.assertRaisesRegex(KeyError, "Key for .images"):
+            _ = self.src.images[3.14]
+
+    def testBadNDArray(self):
+        with self.assertRaisesRegex(
```
Comment (garrettwrong) [inline tests/test_load_images.py:175]: Oh cool, I've not used this method yet. Neat.

**Review by garrettwrong — CHANGES_REQUESTED** [review]
Great, thanks! Couple real minor changes and remarks.  Otherwise I think this is ready for Joakim with the caveat that we need to be really careful about merging this.  It would be very easy to introduce regressions with any PR that interacts with `Sources`.

I expect we will probably need to update this PR after any code merge (or vice versa)

File: src/aspire/source/image.py (hunk 12 above)
Comment (chris-langfield) [reply src/aspire/source/image.py]: That's okay i always nitpick comments and what goes around comes around

File: src/aspire/source/image.py (hunk 11 above)
Comment (chris-langfield) [reply src/aspire/source/image.py]: done

File: src/aspire/source/image.py (hunk 10 above)
Comment (chris-langfield) [reply src/aspire/source/image.py]: done

**Review by garrettwrong — DISMISSED** [review]
I think we're ready to move this along. Thanks!

File: src/aspire/basis/fspca.py [hunk 14]
```diff
@@ -141,7 +141,9 @@ def build(self):
             start = i * self.batch_size
             finish = min((i + 1) * self.batch_size, self.src.n)
             n = finish - start
-            coef[start:finish] = self.basis.evaluate_t(self.src.images(start, n))
+            coef[start:finish] = self.basis.evaluate_t(
```
Comment (janden) [inline src/aspire/basis/fspca.py]: Remove `n` and index `start:finish` instead.

File: tests/test_load_images.py [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+            KeyError, "Only one-dimensional indexing is allowed for images."
+        ):
+            _ = self.src.images[np.zeros((3, 3))]
+
```
Comment (janden) [inline tests/test_load_images.py:179]: Also test for a `slice` input (with non-unit stride, indices outside of range, etc.)?

File: src/aspire/source/image.py [hunk 16]
```diff
@@ -27,7 +29,50 @@
 logger = logging.getLogger(__name__)
 
 
-class ImageSource:
+class ImageAccessor:
```
Comment (janden) [inline src/aspire/source/image.py]: I'm a little wary of pulling this into the public API. One option is to leave it here in the third level (and not import it into `aspire.source.__init__`). Another is to make it private (so `_ImageAccessor`). Latter may be preferable.

File: src/aspire/source/image.py [hunk 17]
```diff
@@ -27,7 +29,50 @@
...
+            if not step:
+                step = 1
+            indices = np.arange(start, stop, step)
+        if not isinstance(indices, np.ndarray):
+            raise KeyError(
+                "Key for .images must be a slice, 1-D NumPy array, or iterable yielding integers."
+            )
+        if not indices.ndim == 1:
+            raise KeyError("Only one-dimensional indexing is allowed for images.")
+        return self.fun(indices)
```
Comment (janden) [inline src/aspire/source/image.py:81]: How do we deal with negative indices (in slice, iterable, or vector form)? Does this need to be coded explicitly? If not, we should probably add tests to make sure that it works correctly.

File: src/aspire/source/image.py [hunk 18]
```diff
@@ -673,7 +702,7 @@ def save_images(
                     logger.info(
                         f"Saving ImageSource[{i_start}-{i_end-1}] to {mrcs_filepath}"
                     )
-                    datum = self.images(start=i_start, num=num).data.astype("float32")
+                    datum = self.images[i_start : i_start + num].data.astype("float32")
```
Comment (janden) [inline src/aspire/source/image.py]: Remove `num` and index `i_star:i_end`. Also below.

File: src/aspire/source/relion.py [hunk 19]
```diff
@@ -227,11 +227,21 @@ def populate_metadata(self, filepath, data_folder=None, max_rows=None):
...
+        Returns particle images when accessed via the `ImageSource.images` property.
+        Loads particle images corresponding to `indices` from StarFile and .mrcs stacks.
+        :param indices: A 1-D NumPy array of integer indices.
+        :return: An `Image` object.
+        """
+
+        # check for cached images first
+        if self._cached_im is not None:
+            logger.info("Loading images from cache")
+            return self.generation_pipeline.forward(
```
Comment (janden) [inline src/aspire/source/relion.py:241]: Seems like an important line to cover in tests.

**Review by janden — CHANGES_REQUESTED** [review]
Looks great! Just a few things.

File: src/aspire/source/image.py (hunk 17 above)
Comment (chris-langfield) [reply src/aspire/source/image.py:81]: This is a potentially larger issue, as the downstream `_images` methods aren't currently set up to handle negative indices.. Looking into it

Comment (chris-langfield) [reply src/aspire/source/image.py:81]: Scratch that, we can convert the negative indices to positive right at the end when the `ImageAccessor` is called, regardless of the type of input

File: src/aspire/basis/fspca.py (hunk 14 above)
Comment (chris-langfield) [reply src/aspire/basis/fspca.py]: Done

File: src/aspire/source/relion.py (hunk 19 above)
Comment (chris-langfield) [reply src/aspire/source/relion.py:241]: Added `cache()` tests for `RelionSource`, `CoordinateSource`, and `Simulation`

File: src/aspire/source/image.py (hunk 18 above)
Comment (chris-langfield) [reply src/aspire/source/image.py]: Done

File: src/aspire/source/image.py (hunk 16 above)
Comment (chris-langfield) [reply src/aspire/source/image.py]: Ended up doing both. We can just import from the third level in the one other place it's used (`aspire/source/simulation.py`) so I'm not sure if the `_` is necessary? I'm fine with either

File: tests/test_load_images.py (hunk 15 above)
Comment (chris-langfield) [reply tests/test_load_images.py:179]: What I did was add a type check for the `slice` components, and then a final check right before calling `_images` to find out of range indices that might have been in `filter`, `range` or `list` arguments.

File: src/aspire/source/image.py (hunk 17 above)
Comment (chris-langfield) [reply src/aspire/source/image.py:81]: Also added tests for negative indices and ranges

**Comment by garrettwrong** [comment]
I fixed up the one minor merge conflict from #696 .

**Review by janden — APPROVED** [review]
